fsharp-backend/src/HttpMiddleware/HttpMiddlewareV1.fs

```diff
@@ -0,0 +1,134 @@
+/// Thin, bytes-friendly middleware
+///
+/// This Middleware corresponds with `HTTPBytes` handlers
+[<RequireQualifiedAccess>]
+module HttpMiddleware.HttpMiddlewareV1
+
+module Request =
+  open Prelude
+  open LibExecution.VendoredTablecloth
+
+  module RT = LibExecution.RuntimeTypes
+  module HttpQueryEncoding = BackendOnlyStdLib.HttpQueryEncoding
+
+  let private parseQueryString (query : string) : RT.Dval =
+    // Drop leading "?"
+    let query = if query.Length > 0 then String.dropLeft 1 query else query
+    HttpQueryEncoding.ofFormEncoding query
+
+  let private parseHeaders (headers : (string * string) list) =
+    headers
+    |> List.map (fun (k, v) -> (String.toLowercase k, RT.DStr v))
+    |> Map
+    |> RT.Dval.DObj
+
+  let private cookies (headers : HttpHeaders.T) : RT.Dval =
+    let parse (cookies : string) : RT.Dval =
+      let decode = System.Web.HttpUtility.UrlDecode
+      cookies
+      |> String.split ";"
+      |> List.map String.trim
+      |> List.map (fun s -> s.Split("=", 2) |> Array.toList)
+      |> List.map (fun cookie ->
+        match cookie with
+        | [] -> ("", RT.DNull) // skip empty rows
+        | [ _ ] -> ("", RT.DNull) // skip rows with only a key
+        | k :: v :: _ -> (decode k, RT.DStr(decode v)))
+      |> RT.Dval.obj
+
+    HttpHeaders.get "cookie" headers
+    |> Option.map parse
+    |> Option.defaultValue (RT.DObj Map.empty)
+
+  let private url (headers : List<string * string>) (uri : string) =
+    // .NET doesn't url-encode the query like we expect, so we're going to do it
+    let parsed = System.UriBuilder(uri)
+    // FSTODO test this somehow (probably fuzz against old)
+    parsed.Query <- urlEncodeExcept "*$@!:()~?/.,&-_=\\" parsed.Query
+    // Set the scheme if it's passed by the load balancer
+    headers
+    |> List.find (fun (k, _) -> String.toLowercase k = "x-forwarded-proto")
+    |> Option.map (fun (_, v) -> parsed.Scheme <- v)
+    |> ignore<Option<unit>>
+    // Use .Uri or it will slip in a port number
+    RT.DStr(string parsed.Uri)
+
+  let fromRequest
+    (uri : string)
+    (headers : List<string * string>)
+    (query : string)
+    (body : byte array)
+    : RT.Dval =
+    [ "body", RT.DBytes body
+      "queryParams", parseQueryString query
```

Comment:
Perhaps the query should be a string instead of parsed?

Reply:
Ah, I didn't think of that somehow - will do!